### PR TITLE
Fix flaky TestConcurrentReads_ReadIsTreatedNonSequentialAfterFileIsRemovedFromCache

### PR DESCRIPTION
**Pull request by ashmeenkaur**
### Description
This PR addresses flaky `TestConcurrentReads_ReadIsTreatedNonSequentialAfterFileIsRemovedFromCache` test, which verifies GCSFuse's cache eviction behavior under concurrent read scenarios. Previously, concurrent reads triggered unpredictable situations like:
1. premature file removal from the cache, even before the initial read operation completed. 
2. logs order inconsistency.
This made the test flaky and unreliable.

To resolve this, the PR modifies the test to perform sequential partial reads one by one. By eliminating concurrent operations, we ensure a deterministic execution flow, allowing the cache to operate as expected. 

### Link to the issue in case of a bug fix.
b/390309943

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: tools/integration_tests/read_cache/helpers_test.go [hunk 1]
```diff
@@ -73,8 +73,8 @@ func readFileAndGetExpectedOutcome(testDirPath, fileName string, readFullFile bo
 	return expected
 }
 
-func validate(expected *Expected, logEntry *read_logs.StructuredReadLogEntry,
-	isSeq, cacheHit bool, chunkCount int, t *testing.T) {
+func validate(expected *Expected, logEntry *read_logs.StructuredReadLogEntry, isSeq, cacheHit bool, chunkCount int, t *testing.T) {
```
Comment (vipnydav) [inline tools/integration_tests/read_cache/helpers_test.go:76]: Shouldn't t *testing.T be first parameter according to Go Style Guide?

File: tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go [hunk 2]
```diff
@@ -86,39 +108,43 @@ func (s *cacheFileForRangeReadFalseTest) TestRangeReadsWithCacheMiss(t *testing.
 	validateFileIsNotCached(testFileName, t)
 }
 
-func (s *cacheFileForRangeReadFalseTest) TestConcurrentReads_ReadIsTreatedNonSequentialAfterFileIsRemovedFromCache(t *testing.T) {
+func (s *cacheFileForRangeReadFalseTest) TestReadIsTreatedNonSequentialAfterFileIsRemovedFromCache(t *testing.T) {
 	var testFileNames [2]string
-	var expectedOutcome [2]*Expected
+	var expectedOutcome [4]*Expected
```
Comment (vipnydav) [inline tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go:113]: Instead of increasing the size to 4, wouldn't it have been better to pass a `WaitGroup` into the `readFileBetweenOffset` method, which is locked twice? This would allow us to use goroutines to read both files concurrently. After reading the first half of each file, we could unlock the lock and wait for it to be fully unlocked before proceeding with the second half. Once the `WaitGroup` is unlocked, the cache would be evicted, as we've read half of both files. This approach would result in cleaner code.

**Review by vipnydav — DISMISSED** [review]
One comment, otherwise LGTM.

File: tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go (hunk 2 above)
Comment (ashmeenkaur) [reply tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go:113]: > Instead of increasing the size to 4, wouldn't it have been better to pass a WaitGroup into the readFileBetweenOffset method, which is locked twice

You are right, this would reduce LOC and the code would look cleaner but I think will hamper the test readability. WDYT?

File: tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go [hunk 3]
```diff
@@ -61,11 +63,31 @@ func (s *cacheFileForRangeReadFalseTest) Teardown(t *testing.T) {
...
+		ObjectName:            path.Join(testDirName, path.Base(file.Name())),
+	}
+	if setup.DynamicBucketMounted() != "" {
+		expected.BucketName = setup.DynamicBucketMounted()
+	}
+
+	chunkRead := make([]byte, chunkSizeToRead)
+	var readData []byte
+	for startOffset < endOffSet {
+		_, err := file.ReadAt(chunkRead, startOffset)
```
Comment (gargnitingoogle) [inline tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go:80]: We are not using endOffSet inside the loop anywhere. If the goal of function is to read from `[startOffset, endOffSet)` of given file, and `endOffset-startOffset` isn't a multiple of `chunkSizeToRead`, then either ReadAt should change itself or we should take a slice of readData later at `readData = append(readData, chunkRead...)` . 

The same goes for `startOffset = startOffset + chunkSizeToRead`, we should increase it only by the actual number of bytes read

Comment (ashmeenkaur) [reply tools/integration_tests/read_cache/cache_file_for_range_read_false_test.go]: Thanks, did some refactoring to make the code generic and reusable. Please take another look!

**Review by vipnydav — DISMISSED** [review]

File: tools/integration_tests/util/operations/file_operations.go [hunk 4]
```diff
@@ -316,6 +316,25 @@ func ReadChunkFromFile(filePath string, chunkSize int64, offset int64, flag int)
...
+func ReadFileBetweenOffset(t *testing.T, file *os.File, startOffset, endOffset, chunkSize int64) string {
+	t.Helper()
+	chunk := make([]byte, chunkSize)
+	var readData []byte
+
+	for startOffset < endOffset {
+		readSize := min(chunkSize, endOffset-startOffset)
+
+		n, err := file.ReadAt(chunk[:readSize], startOffset)
+		if err != nil && err != io.EOF {
```
Comment (gargnitingoogle) [inline tools/integration_tests/util/operations/file_operations.go:328]: This will run into an infinite loop when `endOffser > size` of file, because then error will be EOF and n will be 0 and the loop will keep trying to read the next chunk or remaining bytes. 

We should check here if error is EOF and then terminate based on it. Or, we could fail in the beginning of the function if `endOffser > size`, but requires us to first figure out the file size, which is more expensive.

Comment (gargnitingoogle) [reply tools/integration_tests/util/operations/file_operations.go:328]: this may not affect the test being modified here, but since this function is in a generic util package, its behavior should cater to all boundary cases.

File: tools/integration_tests/util/operations/file_operations.go [hunk 5]
```diff
@@ -325,8 +325,17 @@ func ReadFileBetweenOffset(t *testing.T, file *os.File, startOffset, endOffset,
 		readSize := min(chunkSize, endOffset-startOffset)
 
 		n, err := file.ReadAt(chunk[:readSize], startOffset)
-		if err != nil && err != io.EOF {
-			t.Errorf("Failed to read file chunk at offset %d: %v", startOffset, err)
+		if err != nil {
+			if err == io.EOF && startOffset < endOffset {
```
[the remark this answers is not in the document]
Comment (gargnitingoogle) [reply tools/integration_tests/util/operations/file_operations.go:329]: This also makes the comment below a bit redundant.

File: tools/integration_tests/util/operations/file_operations.go [hunk 6]
```diff
@@ -325,8 +325,17 @@ func ReadFileBetweenOffset(t *testing.T, file *os.File, startOffset, endOffset,
...
+				// Reached EOF before endOffset.
+				if n > 0 {
+					readData = append(readData, chunk[:n]...)
+				}
+				break
+			} else if err != io.EOF {
+				t.Errorf("Failed to read file chunk at offset %d: %v", startOffset, err)
+				return "" //Return empty string on error.
```
Comment (gargnitingoogle) [inline tools/integration_tests/util/operations/file_operations.go:337]: nit: comment seems unnecessary as the code is self-explanatory.

**Review by gargnitingoogle — DISMISSED** [review]
added couple of nit comments about a redundant check and unnecessary comments.

**Review by charith87 — DISMISSED** [review]
